File: tagotip-go/secure.go
package tagotip

import (
	"crypto/aes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
)

const (
	headerSize         = 21
	authHashSize       = 8
	deviceHashSize     = 8
	flagsSize          = 1
	counterSize        = 4
	ccmTagSize         = 8
	ccmNonceSize       = 13
	maxInnerFrameSize  = 16_384
	reservedFlagsValue = 0x41

	flagsCipherMask  = 0b1110_0000
	flagsCipherShift = 5
	flagsVersionMask = 0b0001_1000
	flagsVersionShift = 3
	flagsMethodMask  = 0b0000_0111
)

// CipherSuite represents the AEAD cipher suite.
type CipherSuite int

const (
	CipherSuiteAes128Ccm CipherSuite = 0
)

// EnvelopeMethod represents the method in the envelope flags.
type EnvelopeMethod int

const (
	EnvelopeMethodPush EnvelopeMethod = 0
	EnvelopeMethodPull EnvelopeMethod = 1
	EnvelopeMethodPing EnvelopeMethod = 2
	EnvelopeMethodAck  EnvelopeMethod = 3
)

// EnvelopeHeader represents the parsed 21-byte envelope header.
type EnvelopeHeader struct {
	Flags      byte
	Counter    uint32
	AuthHash   [authHashSize]byte
	DeviceHash [deviceHashSize]byte
}

// SecureError represents an error from crypto envelope operations.
type SecureError struct {
	Message string
}

func (e *SecureError) Error() string {
	return fmt.Sprintf("tagotips: %s", e.Message)
}

func secureErr(msg string) error {
	return &SecureError{Message: msg}
}

// DeriveAuthHash derives the Authorization Hash from a token.
// The token format is "at" + 32 hex chars. The "at" prefix is stripped,
// and SHA-256 is computed over the remaining hex string (UTF-8 encoded).
// Returns the first 8 bytes of the digest.
func DeriveAuthHash(token string) [authHashSize]byte {
	hexPart := token
	if len(token) > 2 && token[0] == 'a' && token[1] == 't' {
		hexPart = token[2:]
	}
	digest := sha256.Sum256([]byte(hexPart))
	var hash [authHashSize]byte
	copy(hash[:], digest[:authHashSize])
	return hash
}

// DeriveDeviceHash derives the Device Hash from a serial number.
// Computes SHA-256 of the serial (UTF-8 encoded) and returns the first 8 bytes.
func DeriveDeviceHash(serial string) [deviceHashSize]byte {
	digest := sha256.Sum256([]byte(serial))
	var hash [deviceHashSize]byte
	copy(hash[:], digest[:deviceHashSize])
	return hash
}

// DeriveKey derives an encryption key from a token and serial using HMAC-SHA256.
// The "at" prefix is stripped from the token. The remaining hex string (UTF-8)
// is used as the HMAC key; the serial (UTF-8) is the HMAC message.
// keyLen must be 16 (AES-128) or 32 (AES-256/ChaCha20).
func DeriveKey(token, serial string, keyLen int) ([]byte, error) {
	if keyLen != 16 && keyLen != 32 {
		return nil, secureErr("key length must be 16 or 32")
	}
	hexPart := token
	if strings.HasPrefix(token, "at") {
		hexPart = token[2:]
	}
	mac := hmac.New(sha256.New, []byte(hexPart))
	mac.Write([]byte(serial))
	fullKey := mac.Sum(nil)
	return fullKey[:keyLen], nil
}

// HexToBytes decodes a hex string into bytes.
func HexToBytes(hexStr string) ([]byte, error) {
	return hex.DecodeString(hexStr)
}

// BytesToHex encodes bytes as a lowercase hex string.
func BytesToHex(data []byte) string {
	return hex.EncodeToString(data)
}

func encodeFlags(cipherID, version, methodID int) (byte, error) {
	flags := byte((cipherID << flagsCipherShift) | (version << flagsVersionShift) | methodID)
	if flags == reservedFlagsValue {
		return 0, secureErr("flags byte 0x41 is reserved")
	}
	return flags, nil
}

func decodeFlags(flags byte) (cipherID, version, methodID int, err error) {
	if flags == reservedFlagsValue {
		return 0, 0, 0, secureErr("flags byte 0x41 is reserved")
	}
	cipherID = int((flags & flagsCipherMask) >> flagsCipherShift)
	version = int((flags & flagsVersionMask) >> flagsVersionShift)
	methodID = int(flags & flagsMethodMask)
	return cipherID, version, methodID, nil
}

func buildEnvelopeHeader(flags byte, counter uint32, authHash [authHashSize]byte, deviceHash [deviceHashSize]byte) []byte {
	header := make([]byte, headerSize)
	header[0] = flags
	binary.BigEndian.PutUint32(header[flagsSize:], counter)
	copy(header[flagsSize+counterSize:], authHash[:])
	copy(header[flagsSize+counterSize+authHashSize:], deviceHash[:])
	return header
}

func constructNonce(flags byte, deviceHash [deviceHashSize]byte, counter uint32) []byte {
	nonce := make([]byte, ccmNonceSize)
	nonce[0] = flags
	// Zero padding at bytes 1-4 (already zeroed)
	// Device hash first 4 bytes at offset (13 - 8) = 5
	copy(nonce[ccmNonceSize-8:ccmNonceSize-4], deviceHash[:4])
	// Counter as big-endian u32 in last 4 bytes
	binary.BigEndian.PutUint32(nonce[ccmNonceSize-4:], counter)
	return nonce
}

// ccmEncrypt performs AES-128-CCM encryption with 8-byte tag.
func ccmEncrypt(key, nonce, aad, plaintext []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, secureErr("invalid encryption key")
	}
	return ccmSeal(block, nonce, aad, plaintext)
}

// ccmDecrypt performs AES-128-CCM decryption with 8-byte tag.
func ccmDecrypt(key, nonce, aad, ciphertextWithTag []byte) ([]byte, error) {
	if len(ciphertextWithTag) < ccmTagSize {
		return nil, secureErr("ciphertext too short")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, secureErr("invalid encryption key")
	}
	return ccmOpen(block, nonce, aad, ciphertextWithTag)
}

// SealUplink encrypts a headless inner frame into a TagoTiP/S uplink envelope.
func SealUplink(
	method EnvelopeMethod,
	innerFrame []byte,
	counter uint32,
	authHash [authHashSize]byte,
	deviceHash [deviceHashSize]byte,
	key []byte,
	suite CipherSuite,
) ([]byte, error) {
	if len(innerFrame) > maxInnerFrameSize {
		return nil, secureErr("inner frame exceeds maximum size")
	}
	if suite != CipherSuiteAes128Ccm {
		return nil, secureErr("unsupported cipher suite")
	}
	if len(key) != 16 {
		return nil, secureErr("invalid encryption key size")
	}

	flags, err := encodeFlags(int(suite), 0, int(method))
	if err != nil {
		return nil, err
	}

	header := buildEnvelopeHeader(flags, counter, authHash, deviceHash)
	nonce := constructNonce(flags, deviceHash, counter)

	ciphertextWithTag, err := ccmEncrypt(key, nonce, header, innerFrame)
	if err != nil {
		return nil, err
	}

	envelope := make([]byte, headerSize+len(ciphertextWithTag))
	copy(envelope, header)
	copy(envelope[headerSize:], ciphertextWithTag)
	return envelope, nil
}

// OpenEnvelope decrypts a TagoTiP/S envelope.
// Returns the header, method, and decrypted inner frame bytes.
func OpenEnvelope(envelope, key []byte) (*EnvelopeHeader, EnvelopeMethod, []byte, error) {
	header, err := ParseEnvelopeHeader(envelope)
	if err != nil {
		return nil, 0, nil, err
	}

	cipherID, version, methodID, err := decodeFlags(header.Flags)
	if err != nil {
		return nil, 0, nil, err
	}

	if version != 0 {
		return nil, 0, nil, secureErr("unsupported version")
	}
	if cipherID != 0 {
		return nil, 0, nil, secureErr("unsupported cipher suite")
	}
	if methodID > 3 {
		return nil, 0, nil, secureErr("invalid method")
	}
	if len(key) != 16 {
		return nil, 0, nil, secureErr("invalid encryption key size")
	}

	ciphertextWithTag := envelope[headerSize:]
	if len(ciphertextWithTag) < ccmTagSize {
		return nil, 0, nil, secureErr("envelope too short")
	}

	aad := envelope[:headerSize]
	nonce := constructNonce(header.Flags, header.DeviceHash, header.Counter)

	plaintext, err := ccmDecrypt(key, nonce, aad, ciphertextWithTag)
	if err != nil {
		return nil, 0, nil, err
	}

	return header, EnvelopeMethod(methodID), plaintext, nil
}

// ParseEnvelopeHeader parses the 21-byte envelope header for server-side routing.
func ParseEnvelopeHeader(envelope []byte) (*EnvelopeHeader, error) {
	if len(envelope) < headerSize {
		return nil, secureErr("envelope too short")
	}

	flags := envelope[0]
	if _, _, _, err := decodeFlags(flags); err != nil {
		return nil, err
	}

	counter := binary.BigEndian.Uint32(envelope[flagsSize:])
	var authHash [authHashSize]byte
	copy(authHash[:], envelope[flagsSize+counterSize:])
	var deviceHash [deviceHashSize]byte
	copy(deviceHash[:], envelope[flagsSize+counterSize+authHashSize:])

	return &EnvelopeHeader{
		Flags:      flags,
		Counter:    counter,
		AuthHash:   authHash,
		DeviceHash: deviceHash,
	}, nil
}

// IsEnvelope checks if a message is a TagoTiP/S envelope or a plaintext fallback.
// Returns true if the first byte is NOT 0x41 (ASCII 'A').
func IsEnvelope(data []byte) bool {
	if len(data) == 0 {
		return false
	}
	return data[0] != reservedFlagsValue
}

// IsSecureError checks if an error is a SecureError.
func IsSecureError(err error) bool {
	var se *SecureError
	return errors.As(err, &se)
}

File: tagotip-node/src/secure.ts
import { createHash, createHmac, createCipheriv, createDecipheriv } from "node:crypto";

const HEADER_SIZE = 21;
const AUTH_HASH_SIZE = 8;
const DEVICE_HASH_SIZE = 8;
const FLAGS_SIZE = 1;
const COUNTER_SIZE = 4;
const CCM_TAG_SIZE = 8;
const CCM_NONCE_SIZE = 13;
const MAX_INNER_FRAME_SIZE = 16_384;
const RESERVED_FLAGS_VALUE = 0x41;

const FLAGS_CIPHER_MASK = 0b1110_0000;
const FLAGS_CIPHER_SHIFT = 5;
const FLAGS_VERSION_MASK = 0b0001_1000;
const FLAGS_VERSION_SHIFT = 3;
const FLAGS_METHOD_MASK = 0b0000_0111;

export type CipherSuite = "aes-128-ccm";

export type EnvelopeMethod = "push" | "pull" | "ping" | "ack";

export interface EnvelopeHeader {
  flags: number;
  counter: number;
  authHash: Uint8Array;
  deviceHash: Uint8Array;
}

export class SecureError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SecureError";
  }
}

const METHOD_IDS: Record<EnvelopeMethod, number> = {
  push: 0,
  pull: 1,
  ping: 2,
  ack: 3,
};

const METHOD_FROM_ID: Record<number, EnvelopeMethod> = {
  0: "push",
  1: "pull",
  2: "ping",
  3: "ack",
};

function sha256(data: Uint8Array): Uint8Array {
  const hash = createHash("sha256");
  hash.update(data);
  return new Uint8Array(hash.digest());
}

export function deriveAuthHash(token: string): Uint8Array {
  const hexPart = token.startsWith("at") ? token.slice(2) : token;
  const digest = sha256(Buffer.from(hexPart, "utf-8"));
  return digest.slice(0, AUTH_HASH_SIZE);
}

export function deriveDeviceHash(serial: string): Uint8Array {
  const digest = sha256(Buffer.from(serial, "utf-8"));
  return digest.slice(0, DEVICE_HASH_SIZE);
}

export function deriveKey(token: string, serial: string, keyLen: 16 | 32 = 16): Uint8Array {
  const hexPart = token.startsWith("at") ? token.slice(2) : token;
  const hmac = createHmac("sha256", Buffer.from(hexPart, "utf-8"));
  hmac.update(Buffer.from(serial, "utf-8"));
  const fullKey = new Uint8Array(hmac.digest());
  return fullKey.slice(0, keyLen);
}

export function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    throw new SecureError("hex string must have even length");
  }
  if (!/^[0-9a-fA-F]*$/.test(hex)) {
    throw new SecureError("hex string contains non-hex characters");
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.substring(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

export function bytesToHex(bytes: Uint8Array): string {
  let hex = "";
  for (let i = 0; i < bytes.length; i++) {
    hex += bytes[i].toString(16).padStart(2, "0");
  }
  return hex;
}

function encodeFlags(cipherSuiteId: number, version: number, methodId: number): number {
  const flags = (cipherSuiteId << FLAGS_CIPHER_SHIFT) | (version << FLAGS_VERSION_SHIFT) | methodId;
  if (flags === RESERVED_FLAGS_VALUE) {
    throw new SecureError("flags byte 0x41 is reserved");
  }
  return flags;
}

function decodeFlags(flags: number): { cipherId: number; version: number; methodId: number } {
  if (flags === RESERVED_FLAGS_VALUE) {
    throw new SecureError("flags byte 0x41 is reserved");
  }
  const cipherId = (flags & FLAGS_CIPHER_MASK) >> FLAGS_CIPHER_SHIFT;
  const version = (flags & FLAGS_VERSION_MASK) >> FLAGS_VERSION_SHIFT;
  const methodId = flags & FLAGS_METHOD_MASK;
  return { cipherId, version, methodId };
}

function buildHeader(flags: number, counter: number, authHash: Uint8Array, deviceHash: Uint8Array): Uint8Array {
  const header = new Uint8Array(HEADER_SIZE);
  header[0] = flags;
  const view = new DataView(header.buffer);
  view.setUint32(FLAGS_SIZE, counter, false);
  header.set(authHash.subarray(0, AUTH_HASH_SIZE), FLAGS_SIZE + COUNTER_SIZE);
  header.set(deviceHash.subarray(0, DEVICE_HASH_SIZE), FLAGS_SIZE + COUNTER_SIZE + AUTH_HASH_SIZE);
  return header;
}

function constructNonce(flags: number, deviceHash: Uint8Array, counter: number): Uint8Array {
  const nonce = new Uint8Array(CCM_NONCE_SIZE);
  nonce[0] = flags;
  // Zero padding at bytes 1-4 (already zeroed)
  // Device hash first 4 bytes at offset 5
  nonce.set(deviceHash.subarray(0, 4), CCM_NONCE_SIZE - 8);
  // Counter as big-endian u32 in last 4 bytes
  const view = new DataView(nonce.buffer);
  view.setUint32(CCM_NONCE_SIZE - 4, counter, false);
  return nonce;
}

function aeadEncrypt(
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array,
  plaintext: Uint8Array,
): Uint8Array {
  const cipher = createCipheriv("aes-128-ccm", key, nonce, { authTagLength: CCM_TAG_SIZE });
  cipher.setAAD(aad, { plaintextLength: plaintext.length });
  const encrypted = cipher.update(plaintext);
  cipher.final();
  const tag = cipher.getAuthTag();
  const result = new Uint8Array(encrypted.length + tag.length);
  result.set(encrypted, 0);
  result.set(tag, encrypted.length);
  return result;
}

function aeadDecrypt(
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array,
  ciphertextWithTag: Uint8Array,
): Uint8Array {
  if (ciphertextWithTag.length < CCM_TAG_SIZE) {
    throw new SecureError("ciphertext too short");
  }
  const ciphertext = ciphertextWithTag.subarray(0, ciphertextWithTag.length - CCM_TAG_SIZE);
  const tag = ciphertextWithTag.subarray(ciphertextWithTag.length - CCM_TAG_SIZE);
  const decipher = createDecipheriv("aes-128-ccm", key, nonce, { authTagLength: CCM_TAG_SIZE });
  decipher.setAuthTag(tag);
  decipher.setAAD(aad, { plaintextLength: ciphertext.length });
  const decrypted = decipher.update(ciphertext);
  try {
    decipher.final();
  } catch {
    throw new SecureError("AEAD decryption failed");
  }
  return new Uint8Array(decrypted);
}

export function sealUplink(
  method: EnvelopeMethod,
  innerFrame: Uint8Array,
  counter: number,
  authHash: Uint8Array,
  deviceHash: Uint8Array,
  key: Uint8Array,
  _suite: CipherSuite = "aes-128-ccm",
): Uint8Array {
  if (innerFrame.length > MAX_INNER_FRAME_SIZE) {
    throw new SecureError("inner frame exceeds maximum size");
  }
  if (key.length !== 16) {
    throw new SecureError("invalid encryption key size");
  }

  const methodId = METHOD_IDS[method];
  if (methodId === undefined) {
    throw new SecureError(`invalid method: ${method}`);
  }
  const flags = encodeFlags(0, 0, methodId);
  const header = buildHeader(flags, counter, authHash, deviceHash);
  const nonce = constructNonce(flags, deviceHash, counter);
  const ciphertextWithTag = aeadEncrypt(key, nonce, header, innerFrame);

  const envelope = new Uint8Array(HEADER_SIZE + ciphertextWithTag.length);
  envelope.set(header, 0);
  envelope.set(ciphertextWithTag, HEADER_SIZE);
  return envelope;
}

export function openEnvelope(
  envelope: Uint8Array,
  key: Uint8Array,
): { header: EnvelopeHeader; method: EnvelopeMethod; plaintext: Uint8Array } {
  const header = parseEnvelopeHeader(envelope);
  const { cipherId, version, methodId } = decodeFlags(header.flags);

  if (version !== 0) {
    throw new SecureError("unsupported version");
  }
  if (cipherId !== 0) {
    throw new SecureError("unsupported cipher suite");
  }
  if (key.length !== 16) {
    throw new SecureError("invalid encryption key size");
  }

  const method = METHOD_FROM_ID[methodId];
  if (method === undefined) {
    throw new SecureError("invalid method");
  }

  const ciphertextWithTag = envelope.subarray(HEADER_SIZE);
  if (ciphertextWithTag.length < CCM_TAG_SIZE) {
    throw new SecureError("envelope too short");
  }

  const aad = envelope.subarray(0, HEADER_SIZE);
  const nonce = constructNonce(header.flags, header.deviceHash, header.counter);
  const plaintext = aeadDecrypt(key, nonce, aad, ciphertextWithTag);

  return { header, method, plaintext };
}

export function parseEnvelopeHeader(envelope: Uint8Array): EnvelopeHeader {
  if (envelope.length < HEADER_SIZE) {
    throw new SecureError("envelope too short");
  }

  const flags = envelope[0];
  decodeFlags(flags);

  const view = new DataView(envelope.buffer, envelope.byteOffset, envelope.byteLength);
  const counter = view.getUint32(FLAGS_SIZE, false);
  const authHash = envelope.slice(FLAGS_SIZE + COUNTER_SIZE, FLAGS_SIZE + COUNTER_SIZE + AUTH_HASH_SIZE);
  const deviceHash = envelope.slice(
    FLAGS_SIZE + COUNTER_SIZE + AUTH_HASH_SIZE,
    FLAGS_SIZE + COUNTER_SIZE + AUTH_HASH_SIZE + DEVICE_HASH_SIZE,
  );

  return { flags, counter, authHash, deviceHash };
}

export function isEnvelope(data: Uint8Array): boolean {
  if (data.length === 0) return false;
  return data[0] !== RESERVED_FLAGS_VALUE;
}
